File: ios/Sources/App/ViewModels/IdentityManagerViewModel.swift
import Foundation
import SwiftData

@MainActor
final class IdentityManagerViewModel: ObservableObject {
    @Published var errorMessage: String?
    @Published var lastGeneratedPrivateKeyPEM: String?

    func generateKey(label: String, type: IdentityKeyType, passphrase: String?, context: ModelContext) {
        do {
            let generated: IdentityKeyGenerator.GeneratedKey
            switch type {
            case .ed25519:
                generated = IdentityKeyGenerator.generateEd25519(comment: label)
            case .rsa4096:
                generated = try IdentityKeyGenerator.generateRSA4096(comment: label)
            }

            let identity = Identity(
                label: label,
                keyType: type,
                fingerprint: generated.fingerprint,
                publicKey: generated.publicKeyLine,
                hasPassphrase: passphrase?.isEmpty == false
            )

            try KeychainService.setPrivateKeyPEM(generated.privateKeyPEM, for: identity)
            if let passphrase, !passphrase.isEmpty {
                try KeychainService.setPassphrase(passphrase, for: identity)
            }

            context.insert(identity)
            lastGeneratedPrivateKeyPEM = generated.privateKeyPEM
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func importKey(
        label: String,
        type: IdentityKeyType,
        privateKeyPEM: String,
        publicKeyLine: String,
        passphrase: String?,
        context: ModelContext
    ) {
        let fingerprint = Self.fingerprint(fromPublicKeyLine: publicKeyLine) ?? "unknown"
        let identity = Identity(
            label: label,
            keyType: type,
            fingerprint: fingerprint,
            publicKey: publicKeyLine,
            hasPassphrase: passphrase?.isEmpty == false
        )
        do {
            try KeychainService.setPrivateKeyPEM(privateKeyPEM, for: identity)
            if let passphrase, !passphrase.isEmpty {
                try KeychainService.setPassphrase(passphrase, for: identity)
            }
            context.insert(identity)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func delete(_ identity: Identity, context: ModelContext) {
        try? KeychainService.deleteSecrets(for: identity)
        context.delete(identity)
    }

    private static func fingerprint(fromPublicKeyLine line: String) -> String? {
        let parts = line.split(separator: " ")
        guard parts.count >= 2, let data = Data(base64Encoded: String(parts[1])) else { return nil }
        return SSHKeyFormat.fingerprint(publicKeyBlob: Array(data))
    }
}

File: android/app/src/main/kotlin/dev/termvault/app/ui/keys/IdentityManagerViewModel.kt
package dev.termvault.app.ui.keys

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dev.termvault.app.data.db.IdentityDao
import dev.termvault.app.data.db.IdentityEntity
import dev.termvault.app.data.model.IdentityKeyType
import dev.termvault.app.security.SecretStore
import dev.termvault.app.ssh.IdentityKeyGenerator
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

class IdentityManagerViewModel(
    private val identityDao: IdentityDao,
    private val secretStore: SecretStore,
) : ViewModel() {
    val identities: StateFlow<List<IdentityEntity>> =
        identityDao.observeAll().stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), emptyList())

    fun generate(label: String, keyType: IdentityKeyType, comment: String) {
        viewModelScope.launch {
            val generated = when (keyType) {
                IdentityKeyType.ED25519 -> IdentityKeyGenerator.generateEd25519(comment)
                IdentityKeyType.RSA4096 -> IdentityKeyGenerator.generateRsa4096(comment)
            }
            val identity = IdentityEntity(
                label = label,
                keyTypeRaw = keyType.rawValue,
                fingerprint = generated.fingerprint,
                publicKey = generated.publicKeyLine,
            )
            identityDao.upsert(identity)
            secretStore.setPrivateKeyPem(generated.privateKeyPem, identity)
        }
    }

    fun delete(identity: IdentityEntity) {
        viewModelScope.launch {
            secretStore.deleteSecrets(identity)
            identityDao.delete(identity)
        }
    }
}
